spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java

```diff
@@ -399,19 +401,52 @@ public synchronized void connectionLost(Throwable cause) {
 
 	@Override
 	public void messageArrived(String topic, MqttMessage mqttMessage) {
-		AbstractIntegrationMessageBuilder<?> builder = getConverter().toMessageBuilder(topic, mqttMessage);
-		if (this.manualAcks) {
-			builder.setHeader(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK,
-					new AcknowledgmentImpl(mqttMessage.getId(), mqttMessage.getQos(), this.client));
+		AbstractIntegrationMessageBuilder<?> builder = toMessageBuilder(topic, mqttMessage);
+		if (builder != null) {
+			if (this.manualAcks) {
+				builder.setHeader(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK,
+						new AcknowledgmentImpl(mqttMessage.getId(), mqttMessage.getQos(), this.client));
+			}
+			Message<?> message = builder.build();
+			try {
+				sendMessage(message);
+			}
+			catch (RuntimeException ex) {
+				logger.error(ex, () -> "Unhandled exception for " + message.toString());
+				throw ex;
+			}
 		}
-		Message<?> message = builder.build();
+	}
+
+	private AbstractIntegrationMessageBuilder<?> toMessageBuilder(String topic, MqttMessage mqttMessage) {
+		AbstractIntegrationMessageBuilder<?> builder = null;
+		Exception conversionError = null;
 		try {
-			sendMessage(message);
+			builder = getConverter().toMessageBuilder(topic, mqttMessage);
 		}
-		catch (RuntimeException ex) {
-			logger.error(ex, () -> "Unhandled exception for " + message.toString());
-			throw ex;
+		catch (Exception ex) {
+			conversionError = ex;
+		}
+
+		if (builder == null && conversionError == null) {
+			conversionError = new IllegalStateException("'MqttMessageConverter' produced a null for message");
+		}
+
+		if (conversionError != null) {
+			GenericMessage<MqttMessage> message = new GenericMessage<>(mqttMessage);
+			if (!sendErrorMessageIfNecessary(message, conversionError)) {
+				MessageConversionException conversionException;
+				if (conversionError instanceof MessageConversionException) {
+					conversionException = (MessageConversionException) conversionError;
+				}
+				else {
+					conversionException = new MessageConversionException(message, "Failed to convert from MQTT Message",
+							conversionError);
+				}
+				throw conversionException;
```

Comment:
Should we still throw here? Or just log as ERROR?

Reply:
Well, I tried to make it consistent with what we have so far in the `MessageProducerSupport.sendMessage()`:
```
try {
	this.messagingTemplate.send(getRequiredOutputChannel(), message);
}
catch (RuntimeException ex) {
	if (!sendErrorMessageIfNecessary(message, ex)) {
		throw ex;
	}
}
```

We probably may consider some general `ErrorHandler` strategy. However it looks like we have it already for this kind of channel adapters via an `ErrorHandlingTaskExecutor` impl. Nevertheless I don't see that we have some way to inject it into the `MqttAsyncClient`...

Probably the matter for a separate issue and discussion over there...

Reply:
Right, but the problem here is that throwing an exception causes the client to close. Conversion exceptions are generally fatal anyway (no point in retrying) since a redelivery will fail again.

However, it would be a behavior change so we'd probably have to add a flag.

Or, we can just leave it as you have it; up to you.

Reply:
Well, we have a flag - `errorChannel` and no exceptions are thrown back to the client!
I would stay with this behavior for a while: probably everybody is going to be happy to swallow and process those wrong messages in the error flow instead of just silent (log) drop to satisfy the client connection stability.